variable.c

```diff
@@ -1478,62 +1507,102 @@ rb_attr_delete(VALUE obj, ID id)
 
-            if (old_ivtbl) {
-                /* We need to modify old_ivtbl to have the too complex shape
+            if (old_fields_tbl) {
+                /* We need to modify old_fields_tbl to have the too complex shape
                  * and hold the table because the xmalloc could trigger a GC
                  * compaction. We want the table to be updated rather than
-                 * the original ivptr. */
+                 * the original fields. */
 #if SHAPE_IN_BASIC_FLAGS
-                rb_shape_set_shape_id(obj, OBJ_TOO_COMPLEX_SHAPE_ID);
+                rb_shape_set_shape_id(obj, shape_id);
 #else
-                old_ivtbl->shape_id = OBJ_TOO_COMPLEX_SHAPE_ID;
+                old_fields_tbl->shape_id = shape_id;
 #endif
-                old_ivtbl->as.complex.table = table;
-                old_ivptr = (VALUE *)old_ivtbl;
+                old_fields_tbl->as.complex.table = table;
+                old_fields = (VALUE *)old_fields_tbl;
             }
 
-            struct gen_ivtbl *ivtbl = xmalloc(sizeof(struct gen_ivtbl));
-            ivtbl->as.complex.table = table;
-            st_insert(gen_ivs, (st_data_t)obj, (st_data_t)ivtbl);
+            struct gen_fields_tbl *fields_tbl = xmalloc(sizeof(struct gen_fields_tbl));
+            fields_tbl->as.complex.table = table;
+            st_insert(gen_ivs, (st_data_t)obj, (st_data_t)fields_tbl);
+
 #if SHAPE_IN_BASIC_FLAGS
-            rb_shape_set_shape_id(obj, OBJ_TOO_COMPLEX_SHAPE_ID);
+            rb_shape_set_shape_id(obj, shape_id);
 #else
-            ivtbl->shape_id = OBJ_TOO_COMPLEX_SHAPE_ID;
+            fields_tbl->shape_id = shape_id;
 #endif
         }
         RB_VM_LOCK_LEAVE();
     }
 
-    xfree(old_ivptr);
+    xfree(old_fields);
+}
+
+void
+rb_obj_init_too_complex(VALUE obj, st_table *table)
+{
+    // This method is meant to be called on newly allocated object.
+    RUBY_ASSERT(!rb_shape_obj_too_complex(obj));
+    RUBY_ASSERT(rb_shape_canonical_p(rb_shape_get_shape(obj)));
+    RUBY_ASSERT(rb_shape_get_shape(obj)->next_field_index == 0);
+
+    obj_transition_too_complex(obj, table);
+}
+
+// Copy all object fields, including ivars and internal object_id, etc
+void
+rb_evict_fields_to_hash(VALUE obj)
+{
+    void rb_obj_copy_fields_to_hash_table(VALUE obj, st_table *table);
+
+    RUBY_ASSERT(!rb_shape_obj_too_complex(obj));
+
+    st_table *table = st_init_numtable_with_size(rb_ivar_count(obj));
+
+    // Evacuate all previous values from shape into id_table
+    rb_obj_copy_fields_to_hash_table(obj, table);
+    rb_shape_t *shape = rb_shape_get_shape(obj);
+    if (rb_shape_has_object_id(shape)) {
+        rb_shape_t *object_id_shape = rb_shape_object_id_shape(obj);
+        VALUE id = rb_field_get(obj, object_id_shape);
+        st_insert(table, internal_object_id, id);
+
+        // We need to ensure the object ID is registered in id_to_obj_table
+        // before transitioning to too complex
+        rb_gc_update_id_to_obj_table(obj, id);
+    }
```

Comment:
This part I don't get.

It was added by @eightbitraptor in https://github.com/byroot/ruby/commit/1023700f019. To me it sounds like useless code because `rb_obj_copy_fields_to_hash_table` should already copy the `object_id` in the `st_table`.

But if I remove this code CI immediately fail, particularly with mmtk, so I assume there's some sort of race condition here.

Reply:
We investigated this.

What seem to be going on:

  - A `T_DATA` or `T_FILE` enter `rb_gc_obj_free_vm_weak_references`.
  - Its shape indicate it has an `object_id`, but when accessing it, we get garbage.
  - That cause the `object_id_hash` function to crash, because it expect either `T_FIXNUM` or `T_BIGNUM` and what we get is neither (we observed a `T_FILE`).

We suspect `rb_gc_obj_free_vm_weak_references` is sometimes called twice on some objects. Before it wasn't a problem because `rb_gc_obj_free_vm_weak_references` was idempotent, but with this PR it no longer is.

Question:

  - Is it actually a contract that it must be idempotent? Or just an happy accident?
  - Is it expected that an object may be freed twice?

Reply:
I figured that one out: 61fb42169693050a8a42bbff32d2ac8b89a1355f

What was happening is I made the mistake of declaring `internal_object_id` as a `static` in `shape.h`, so it was only initialized on `shape.c`, everywhere else it was either `0` or garbage. So this code was preventing the bug from trigerring by aliasing the `object_id` key with whatever the garbage ID was.